File: javascript/75.sort-colors.js
/*
 * @lc app=leetcode id=75 lang=javascript
 *
 * [75] Sort Colors
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function(nums) {
  // Bucket sort
  // O(n) time
  // O(n) space
  let buckets = {}
  for (let num of nums) {
    buckets[num] ? buckets[num] += 1 : buckets[num] = 1
  }
  let index = 0
  for (let [k, v] of Object.entries(buckets)) {
    for (let i = 0; i < v; i++) {
      nums[index] = parseInt(k)
      index++
    }
  }
  return nums
};
// @lc code=end

console.log(sortColors([2,0,2,1,1,0]))

File: cpp/75.sort-colors.cpp
/*
 * @lc app=leetcode id=75 lang=cpp
 *
 * [75] Sort Colors
 */

// @lc code=start
class Solution
{
public:
    void sortColors(vector<int> &nums)
    {
        // bucket sort variant
        // count each num
        // overwrite nums with counters
        // O(n) time O(k) space
        int zeroCount = 0;
        int oneCount = 0;
        for (int i = 0; i < nums.size(); i++)
        {
            switch (nums[i])
            {
            case 0:
                zeroCount++;
                break;
            case 1:
                oneCount++;
                break;
            }
        }
        for (int i = 0; i < nums.size(); i++)
        {
            if (zeroCount > 0)
            {
                nums[i] = 0;
                zeroCount--;
            }
            else if (oneCount > 0)
            {
                nums[i] = 1;
                oneCount--;
            }
            else
            {
                nums[i] = 2;
            }
        }
    }
};
// @lc code=end
